src/emlid_reach.cpp

```diff
@@ -0,0 +1,434 @@
+					_parse_err_cnt ++;
+				}
+			}
+		}
+	}
+}
+
+
+//// ERB
+
+int
+GPSDriverEmlidReach::erbParseChar(uint8_t b)
+{
+	int ret = 0;
+
+	if (_erb_decode_state != ERB_State::init && b == ERB_SYNC_1) {
+		// got SYNC byte when not expected, restart parser
+		_buff_len = 0;
+		_buff[_buff_len ++] = b;
+		_erb_decode_state = ERB_State::got_sync_1;
+		_parse_err_cnt ++;
+		return 0;
+	}
+
+	switch (_erb_decode_state) {
+	case ERB_State::init:
+		if (b == ERB_SYNC_1) {
+			_buff_len = 0;
+			_buff[_buff_len ++] = b;
+			_erb_decode_state = ERB_State::got_sync_1;
+		}
+
+		break;
+
+	case ERB_State::got_sync_1:
+		if (b == ERB_SYNC_2) {
+			_buff[_buff_len ++] = b;
+			_erb_decode_state = ERB_State::got_sync_2;
+
+		} else {
+			_erb_decode_state = ERB_State::init;
+			_parse_err_cnt ++;
+		}
+
+		break;
+
+	case ERB_State::got_sync_2:
+		if (b >= ERB_ID_VERSION && b <= ERB_ID_RTK) {
+			_buff[_buff_len ++] = b;
+			_erb_decode_state = ERB_State::got_id;
+
+			if (b == ERB_ID_SPACE_INFO || b == ERB_ID_RTK) {
+				//ignore those
+				_erb_decode_state = ERB_State::init;
+			}
+
+		} else {
+			_erb_decode_state = ERB_State::init;
+			_parse_err_cnt ++;
+		}
+
+		break;
+
+	case ERB_State::got_id:
+		_buff[_buff_len ++] = b;
+		_erb_decode_state = ERB_State::got_len_1;
+		_erb_payload_len = b;
+		break;
+
+	case ERB_State::got_len_1:
+		_buff[_buff_len ++] = b;
+		_erb_decode_state = ERB_State::got_len_2;
+		_erb_payload_len = (b << 8) | _erb_payload_len;
+		break;
+
+	case ERB_State::got_len_2:
+		if (_buff_len >= ERB_SENTENCE_MAX_LEN) {
+			_erb_decode_state = ERB_State::init;
+			_parse_err_cnt ++;
+			//GPS_INFO("EMLIDREACH: ERB payload overflow");
+
+		} else {
+			_buff[_buff_len ++] = b;
+		}
+
+		if (_buff_len >= _erb_payload_len + static_cast<unsigned>(ERB_HEADER_LEN)) {
+			_erb_decode_state = ERB_State::got_payload;
+		}
+
+		break;
+
+	case ERB_State::got_payload:
+		_checksum_buff_len = 0;
+		_checksum_buff[_checksum_buff_len ++] = b;
+		_erb_decode_state = ERB_State::got_CK_A;
+		break;
+
+	case ERB_State::got_CK_A:
+		_checksum_buff[_checksum_buff_len ++] = b;
+
+		uint8_t cka = 0, ckb = 0;
+
+		for (unsigned i = 2; i < _erb_payload_len + static_cast<unsigned>(ERB_HEADER_LEN); i++) {
+			cka += _buff[i];
+			ckb += cka;
+		}
+
+		if (cka == _checksum_buff[0] && ckb == _checksum_buff[1]) {
+			ret = 1;
+
+		} else {
+			//GPS_INFO("EMLIDREACH: ERB checksum NOK");
+			ret = 0;
+			_parse_err_cnt ++;
+		}
+
+		_erb_decode_state = ERB_State::init;
+		break;
+	}
+
+	return ret;
+}
+
+int
+GPSDriverEmlidReach::handleErbSentence()
+{
+	int ret = 0;
+
+	if (_buff[2] == ERB_ID_VERSION) {
+		//GPS_INFO("EMLIDREACH: ERB version: %d.%d.%d", _buff[ERB_HEADER_LEN + 4], _buff[ERB_HEADER_LEN + 5], _buff[ERB_HEADER_LEN + 6]);
+
+	} else if (_buff[2] == ERB_ID_GEODIC_POSITION) {
+
+		double tmp_double = 0.0;
+		_gps_position->timestamp = gps_absolute_time();
+
+		memcpy(&_last_POS_timeGPS, _buff + ERB_HEADER_LEN, sizeof(uint32_t));
```

Comment:
If you created a union of (packed) structs for `_buff`, you could avoid the memcpy and offsets, and the code would be more readable.
For example as the u-blox driver does: https://github.com/PX4/GpsDrivers/blob/c8fb76fdeb80a6a579bf0ba303a78f60cc40835c/src/ubx.h#L706.

Reply:
doing that